[#171] Feat: 프로젝트 랭킹 스냅샷 응답의 조회 성능 개선을 위한 Redis 캐시 적용

juintination opened the pull request:
## ⭐ Key Changes

1. Redis 캐시 도입 (`RedisTemplate<String, Object>` 활용)
2. 스냅샷 ID 기반 캐시 키 구성 (e.g., `ranking:snapshot:{id}`)
3. 캐시가 존재할 경우 DB 조회 생략
4. 캐시 미스 시 DB 조회 후 Redis에 저장 (TTL: configurable)
5. 관련 설정 값 `.yml` 환경 변수화

<br />

## 🖐️ To reviewers

1. TTL 설정값 및 키 구성 방식의 명확성과 일관성 검토 부탁드립니다.
2. 캐시 무효화 및 최신성 보장을 위한 현재 로직의 설계에 개선 여지가 있는지 검토해주시면 감사하겠습니다.

<br />

## 📌 issue

- close #171

juintination committed:
feat: 프로젝트 랭킹 스냅샷 Redis 캐싱 기능 추가

- ProjectRankingSnapshotServiceImpl에 RedisTemplate 연동
- 스냅샷 응답 DTO(ProjectRankingSnapshotResponseDTO)를 Redis에 저장/조회하도록 변경
- 캐시 key-prefix를 환경 변수(ranking.snapshot.cache.key-prefix)로 분리하여 유연성 확보
- snapshot 생성 시에도 캐시에 즉시 저장되도록 개선

juintination committed:
test: Redis 캐시 기능에 대한 통합 테스트 코드 추가

- RedisTemplate을 활용한 기본 캐시 동작에 대한 테스트 시나리오 구현
  - 데이터 저장 후 조회 성공 케이스
  - TTL(Time To Live) 만료 후 캐시 삭제 확인
  - 존재하지 않는 키 조회 시 null 반환 확인
- 테스트 코드의 중복 제거를 위한 RedisKeyValueFixture 클래스 작성
- 테스트 환경에서 Redis 정상 동작 여부를 보장하기 위한 기초 테스트 구성

juintination committed:
refactor: Redis 기반 프로젝트 생성 시간 캐싱 로직 도입 및 스냅샷 생성 최적화

- 최근 생성된 프로젝트의 createdAt을 Redis에 저장하여 매번 DB 조회 없이 스냅샷 갱신 여부 판단
- snapshot 생성 시 Redis 키 prefix와 latestCreatedAt 키를 환경변수로 관리하도록 개선
- ProjectRepository에 findLatestCreatedAt() 쿼리 메서드 추가
- ProjectRankingSnapshotServiceImpl의 register 로직 내 Redis 캐시 우선 조회 방식 적용

juintination committed:
test: Redis 캐시 관련 통합 테스트 추가

- TTL 만료 후 동일 키로 값 덮어쓰기 시 TTL이 정상적으로 갱신되는지 검증
- 덮어쓰기 후 TTL 내에 캐시가 유지되는지 확인

juintination committed:
test: TTL 및 Sleep 값을 외부 주입 가능하도록 Redis 통합 테스트 개선

- @TestPropertySource를 사용하여 TTL(Duration)과 Sleep 간격을 속성으로 주입
- 테스트 실행 시간을 단축하고 유연한 TTL 테스트 가능
- TTL 및 갱신 동작 테스트에서 고정값 대신 주입값 사용으로 확장성 강화

juintination committed:
refactor: 최근 프로젝트 생성 시간 조회 로직 분리 및 리팩토링

- ProjectRecencyService를 도입하여 최신 프로젝트 생성 시간 조회 책임을 위임
- ProjectRankingSnapshotServiceImpl에서 직접 Repository 접근 대신 RecencyService를 통해 추상화된 로직 사용
- 코드 가독성과 테스트 용이성 향상

juintination committed:
Merge branch 'dev' into optimize/ranking-cache-redis

juintination committed:
refactor: 프로젝트 랭킹 스냅샷 동시성 제어 및 캐싱 로직 개선

- Redisson 기반 분산락으로 중복 스냅샷 생성 방지
- Redis 캐시를 통해 일정 시간 내 스냅샷 ID 재사용
- 캐시 키 및 락 키를 환경변수로 외부 설정 가능하도록 분리
- 동시성 상황을 검증하는 통합 테스트 추가

juintination committed:
feat: TTL 기반 캐시와 DB fallback 병행으로 스냅샷 중복 생성 방지 강화

- 캐시 TTL 만료로 인한 중복 생성 방지를 위해 DB 기반 fallback 로직 추가
- 최신 스냅샷 캐시가 없을 경우 DB에서 유효한 스냅샷 판단 후 재사용 처리
- projectLatestCreatedAtKey 캐시 제거 및 관련 로직 정리

juintination committed:
refactor: 프로젝트 랭킹 스냅샷 생성 시 DTO 기반 JSON 직렬화 방식으로 개선

- Map<String, Object> 기반 JSON 생성 방식 제거
- RankingItemDTO 리스트를 직접 구성해 ObjectMapper로 직렬화
- 가독성과 타입 안정성 향상

juintination committed:
refactor: Redis 플래그 키를 활용한 스냅샷 중복 호출 방지 로직 추가

- snapshot:generating 키를 기반으로 중복 스냅샷 등록 방지 (SETNX 방식 적용)
- TTL 만료 전 중복 요청 발생 시 예외 처리
- 스냅샷 생성 완료 후 플래그 키 제거하여 정상 흐름 복원
- fallback 처리로 TTL 소멸 시에도 DB 기반 스냅샷 재사용 가능

juintination committed:
refactor: ProjectRankingSnapshot 도메인 전반에 BusinessException 적용

- ProjectRankingSnapshotServiceImpl 내부 예외를 BusinessException으로 통일
  - 분산 락 획득 실패: SNAPSHOT_LOCK_UNAVAILABLE
  - 인터럽트 발생 시: SNAPSHOT_LOCK_INTERRUPTED
  - JSON 직렬화 실패 시: SNAPSHOT_SERIALIZATION_FAILED
- 하드코딩된 에러 메시지를 Enum 기반 구조로 일원화하여 유지보수성과 일관성 개선

juintination committed:
refactor: 스냅샷 중복 생성 예외 BusinessException 적용

juintination committed:
refactor: ProjectRankingSnapshot 생성 로직 메서드 분리

- createAndSaveSnapshot() 메서드를 책임 단위로 분리
  - generateRanking(): 프로젝트 랭킹 리스트 구성
  - serializeToJson(): JSON 직렬화 처리
  - persistSnapshot(): 스냅샷 DB 저장 및 flush
  - cacheSnapshot(): Redis에 스냅샷 캐시 저장
- 단일 책임 원칙(SRP)을 적용하여 가독성 및 유지보수성 향상

2sg reviewed (commented):
지금은 TTL로만 캐시 무효화를 하는게 맞을까요? 의도한 내용인지 궁금합니다.

juintination committed:
feat: ProjectPagingRepository에 Redis 캐시 적용

- 스냅샷 기반 프로젝트 랭킹 페이지네이션에서 Redis 캐시를 활용하도록 수정
- Redis에 캐시된 ProjectRankingSnapshotResponseDTO가 존재하면 DB 조회 없이 사용
- 캐시 미존재 시 기존 스냅샷 데이터를 파싱해 사용

juintination committed:
refactor: Redis 캐시 조회 시 명시적 StringRedisTemplate 사용으로 타입 안정성 강화

- 스냅샷 ID 캐시를 조회할 때 Object → String 캐스팅 대신 RedisTemplate<String, String> 사용
- RedisConfig에 stringRedisTemplate 빈 추가
- 캐스팅 오류(ClassCastException) 방지 및 명시적 타입 사용으로 가독성과 안전성 향상

juintination commented on src/main/java/com/tebutebu/apiserver/service/project/snapshot/ProjectRankingSnapshotServiceImpl.java:
```diff
@@ -29,61 +36,168 @@ public class ProjectRankingSnapshotServiceImpl implements ProjectRankingSnapshot
+
+            // 캐시에 저장된 스냅샷 ID가 있는 경우 재사용
+            String latestCacheKey = snapshotCacheKeyPrefix + snapshotCacheKeyLatestSuffix;
+            String cachedSnapshotId = (String) redisTemplate.opsForValue().get(latestCacheKey);
```
Object를 String으로 캐스팅하는 부분에서 타입 불일치 시 ClassCastException이 발생할 수 있는 여지가 있어, RedisTemplate<String, String>을 별도로 사용하도록 수정하여 타입 안정성을 확보했습니다.

juintination committed:
refactor: 예외 발생 시 snapshot generating 키가 삭제되지 않도록 조건부 삭제 적용

- 스냅샷 생성 도중 예외가 발생해도 generating 키가 삭제되던 로직 수정
- 생성이 성공한 경우에만 generating 키 삭제하도록 success 플래그 도입
- 중복 생성 방지 및 Redis TTL 보존을 통해 안정성 강화

juintination commented on src/main/java/com/tebutebu/apiserver/service/project/snapshot/ProjectRankingSnapshotServiceImpl.java:
```diff
@@ -29,61 +36,168 @@ public class ProjectRankingSnapshotServiceImpl implements ProjectRankingSnapshot
+                lock.unlock();
+                log.info("Lock released for snapshot registration.");
+            }
+        }
```
예외 발생 시에도 generating 키가 삭제되어 중복 생성 위험이 있었습니다. 성공 시에만 삭제되도록 조건 분기 로직을 추가하여 문제를 해결했습니다.

juintination committed:
chore: RedisTemplate<Boolean> 빈 추가 및 Object 기반 템플릿 제거

- 스냅샷 생성 중 여부 캐싱을 위한 Boolean 타입 템플릿 추가
- RedisTemplate<String, Object> 제거하여 타입별 RedisTemplate 명시화
- 불필요한 RedisTemplate<Object> 통합 테스트 제거

juintination committed:
fix: snapshot JSON 역직렬화 오류 수정 및 Redis 캐싱 로직 정리

- snapshotRedisTemplate은 DTO 객체를 직렬화/역직렬화하도록 GenericJackson2JsonRedisSerializer 설정
- ProjectRankingSnapshotService.entityToDTO 메서드에서 이중 직렬화된 JSON 문자열 처리 로직 추가
- ProjectRankingSnapshotServiceImpl.cacheSnapshot 시 Redis에 저장하는 값은 DTO 객체 그대로 사용하도록 수정
- snapshotGeneratingKey 삭제 시 booleanRedisTemplate만 사용

juintination committed:
test: getLatestSnapshot 캐싱 로직 검증을 위한 테스트 추가

- Redis에 Snapshot ID만 존재하고 DTO 본문 캐시가 없는 경우, DB 조회 후 캐싱 동작 검증
- DTO가 캐시에 존재할 경우 DB 조회 없이 Redis 캐시만으로 응답이 반환되는지 확인

juintination committed:
refactor: 락 대기 및 점유 시간 하드코딩 제거 및 의미 있는 변수로 추출

- lock.tryLock()에 하드코딩되어 있던 대기 시간(15초)과 점유 시간(60초)을 의미 있는 변수 (`lockWaitTimeSeconds`, `lockLeaseTimeSeconds`)로 분리
- 주석 없이도 코드의 의도가 명확히 드러나도록 개선

juintination commented on src/main/resources/application-test.yml:
```diff
@@ -89,6 +96,13 @@ ranking:
+    cache:
+      key-prefix: "ranking:snapshot:"
+      key-latest-suffix: latest:id
+      key-generating-flag: snapshot:generating
```
혹시 `key-generating-flag` 관련해서 앞에 `ranking:`이 붙어야 할 것 같다는 말씀이신가요?

juintination committed:
test: ProjectRankingSnapshot 동시성 테스트 개선 및 결과 형식 검증 추가

- `entityToDTO()` 메서드를 활용하여 결과 JSON 역직렬화 처리 일관성 확보
- 생성된 Snapshot의 rank 정렬 여부 및 데이터 형식까지 검증

juintination commented on src/test/java/com/tebutebu/apiserver/integration/concurrency/project/snapshot/ProjectRankingSnapshotConcurrencyTest.java:
```diff
@@ -0,0 +1,119 @@
+@SpringBootTest
+@ActiveProfiles("test")
+@DisplayName("ProjectRankingSnapshot 동시성 테스트")
+public class ProjectRankingSnapshotConcurrencyTest {
```
해당 피드백을 반영하여 concurrentRegister_createsSingleSnapshot() 테스트에 다음과 같은 형식 검증 로직을 추가했습니다:

- entityToDTO()를 활용하여 JSON 파싱 오류 여부를 검증하고,
- data 리스트가 비어있지 않은 경우:
  - 첫 번째 항목의 rank가 1인지 확인하고,
  - 전체 리스트가 rank 기준으로 정렬되어 있는지를 추가 검증하였습니다.

이를 통해 단순히 Snapshot ID만 생성되는 것이 아니라, 정상적인 구조의 랭킹 데이터가 포함되어 있는지까지 검증할 수 있도록 보완하였습니다.

juintination commented on src/test/java/com/tebutebu/apiserver/integration/concurrency/project/snapshot/ProjectRankingSnapshotConcurrencyTest.java:
```diff
@@ -0,0 +1,119 @@
+            assertThat(exceptions).isEmpty();
+            assertThat(snapshotIds).hasSize(threadCount);
+            assertThat(distinctIds).hasSize(1);
+            assertThat(snapshotRepository.count()).isEqualTo(1);
```
해당 피드백을 반영하여 concurrentRegister_createsSingleSnapshot() 테스트에 다음과 같은 형식 검증 로직을 추가했습니다:

- entityToDTO()를 활용하여 JSON 파싱 오류 여부를 검증하고,
- data 리스트가 비어있지 않은 경우:
  - 첫 번째 항목의 rank가 1인지 확인하고,
  - 전체 리스트가 rank 기준으로 정렬되어 있는지를 추가 검증하였습니다.

이를 통해 단순히 Snapshot ID만 생성되는 것이 아니라, 정상적인 구조의 랭킹 데이터가 포함되어 있는지까지 검증할 수 있도록 보완하였습니다.

juintination committed:
fix: Redis 캐시 미존재 시 DB 스냅샷 fallback 로직 개선

- Redis에 snapshot 캐시가 없을 경우 DB에서 최근 스냅샷을 조회하여 재사용 가능하도록 수정
- 중복 생성 방지 플래그(redis key: snapshot:generating) 확인 시점을 뒤로 미뤄, DB fallback 후에도 예외 없이 처리되도록 개선
- Redis 장애 상황 또는 TTL 만료 시에도 graceful 하게 동작

juintination committed:
Merge branch 'dev' into optimize/ranking-cache-redis

juintination committed:
fix: 프로젝트 랭킹별 페이지네이션 로직 관련 버그 수정

- 이전에 수정된 start 계산 관련 로직 롤백

juintination commented on src/main/java/com/tebutebu/apiserver/service/project/snapshot/ProjectRankingSnapshotServiceImpl.java:
```diff
@@ -29,61 +36,168 @@ public class ProjectRankingSnapshotServiceImpl implements ProjectRankingSnapshot
         }
+    }
 
+    private ProjectRankingSnapshot persistSnapshot(String json) {
```
register() 메서드는 아래와 같이 인터페이스에 @Transactional이 선언되어 있습니다.
```
@Transactional
public interface ProjectRankingSnapshotService {

    Long register();

    /* 생략 */

}
```
persistSnapshot()은 createAndSaveSnapshot()을 통해 register() 내부에서 호출되므로, 트랜잭션 범위 내에서 실행됩니다.
따라서 persistSnapshot() 내에서 flush()를 호출하더라도 트랜잭션은 정상적으로 적용되며, 별도로 @Transactional을 추가할 필요는 없습니다.

juintination committed:
refactor: flatten snapshot fallback 조건문으로 가독성 개선

juintination committed:
refactor: register() 메서드 책임 분리로 가독성 및 유지보수성 개선

juintination commented on src/main/java/com/tebutebu/apiserver/service/project/snapshot/ProjectRankingSnapshotServiceImpl.java:
```diff
@@ -29,61 +36,168 @@ public class ProjectRankingSnapshotServiceImpl implements ProjectRankingSnapshot
+    private long snapshotGeneratingTtlSeconds;
+
     @Override
     public Long register() {
```
피드백 내용에 따라 register() 메서드의 책임을 다음과 같이 명확히 분리하여 단일 책임 원칙을 적용하고 가독성 및 유지보수성을 개선했습니다.
- 락 획득 및 해제 로직 분리: acquireLock()과 releaseLock() 메서드로 분리하여 Redisson 락 처리의 책임을 명확히 함
- 캐시 스냅샷 ID 재사용 로직 분리: getSnapshotIdFromCache() 메서드를 통해 Redis 캐시에 존재하는 스냅샷 ID 조회 및 재사용 여부를 판단
- DB 기반 fallback 스냅샷 재사용 로직 분리: getSnapshotIdFromDbFallback() 메서드로 최근 DB 스냅샷을 재활용할 수 있는지 판단 및 캐싱 처리
- 중복 생성 방지 플래그 확인 로직 분리: checkIfGeneratingInProgress() 메서드에서 중복 요청 여부를 제어
- 스냅샷 생성 로직 유지: createAndSaveSnapshot()은 기존처럼 분리된 로직으로 스냅샷 생성 담당

juintination commented on src/test/java/com/tebutebu/apiserver/integration/concurrency/project/snapshot/ProjectRankingSnapshotConcurrencyTest.java:
```diff
@@ -0,0 +1,119 @@
+    @BeforeEach
+    void clearSnapshots() {
+        snapshotRepository.deleteAll();
+        // Clear Redis cache as well
```
해당 주석은 삭제하였습니다!

Mr-Muji committed:
feat: 이미지 태그 설정 변경( "/" 대신 "-" 사용)

juintination committed:
fix: 인증이 필요한 Redis 서버 연결 시 비밀번호 설정 추가

- spring.data.redis.password 값을 기반으로 RedissonClient에 비밀번호 설정
- 비밀번호가 설정된 Redis 서버에서 발생하던 NOAUTH 예외 해결

juintination committed:
fix: snapshot 캐시 조회 시 cachedSnapshotId null 방어 처리

juintination committed:
Merge branch 'dev' into optimize/ranking-cache-redis

juintination committed:
Merge branch 'dev' into optimize/ranking-cache-redis
